crhchttp/config.go

```diff
@@ -6,19 +6,96 @@
 package crhchttp
 
 import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	neturl "net/url"
+
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	metricscfgv1beta1 "github.com/project-koku/koku-metrics-operator/api/v1beta1"
 )
 
 // AuthConfig provides the data for reconciling the CR with defaults
 type AuthConfig struct {
-	Client            client.Client
-	ClusterID         string
-	Authentication    metricscfgv1beta1.AuthenticationType
-	BearerTokenString string
-	BasicAuthUser     string
-	BasicAuthPassword string
-	ValidateCert      bool
-	OperatorCommit    string
+	Client              client.Client
+	ClusterID           string
+	Authentication      metricscfgv1beta1.AuthenticationType
+	BearerTokenString   string
+	BasicAuthUser       string
+	BasicAuthPassword   string
+	ValidateCert        bool
+	OperatorCommit      string
+	ServiceAccountData  ServiceAccountData
+	ServiceAccountToken ServiceAccountToken
+}
+
+// ServiceAccountData provides the data for acquiring the service acount token
+type ServiceAccountData struct {
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+	GrantType    string `json:"grant_type"`
+}
+type ServiceAccountToken struct {
+	AccessToken      string `json:"access_token"`
+	ExpiresIn        int    `json:"expires_in"`
+	RefreshExpiresIn int    `json:"refresh_expires_in"`
+	TokenType        string `json:"token_type"`
+	NotBeforePolicy  int    `json:"not_before_policy"`
+	Scope            string `json:"scope"`
+}
+
+const serviceaccount = metricscfgv1beta1.ServiceAccount
+
+func (ac *AuthConfig) GetAccessToken(tokenURL string) error {
+	if ac.Authentication != serviceaccount {
+		return nil
+	}
+
+	log := log.WithName("GetAccessToken")
+
+	// Prepare the POST data
+	data := neturl.Values{}
+	data.Set("client_id", ac.ServiceAccountData.ClientID)
+	data.Set("client_secret", ac.ServiceAccountData.ClientSecret)
+	data.Set("grant_type", ac.ServiceAccountData.GrantType)
+
+	// // Making the HTTP POST request
+	resp, err := http.Post(tokenURL, "application/x-www-form-urlencoded", bytes.NewBufferString(data.Encode()))
```

Comment:
After thinking about this a little bit, I think we need to use `context`, and specify a timeout (just in case).

That requires using `http.NewRequestWithContext` and a `client.Do`.

```
ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second) <<<< or even better to pass `ctx` into `GetAccessToken`, and use that here
defer cancel()
req, err := http.NewRequestWithContext(ctx, http.MethodPost, tokenURL, bytes.NewBufferString(data.Encode()))
// handle err
```
and then:
```
resp, err := http.DefaultClient.Do(req)
```
It IS a little annoying to handle yet another error with `newrequestwithcontext`, but still good to do regardless.

Reply:
absolutely, I've updated request to add a timeout